### PR TITLE
Spark-Mesos plugin implementation

### PR DESCRIPTION
**Pull request by betobetodev**
This merge request contains the full implementation and changes needed to run an Spark application over Mesos in an infrastructure provided by OpenNebula. There are some points that are very specific to work with the infrastructure security limitations that we have in the context of BigSea EuBra partners. It means that this plugin is not completely generic, because it works following the GryCAP network topology and security policies. But this is a great base to write any other plugin to do similar tasks.

This plugin submit (via ssh) an spark application with the input and output stored in some accessible HDFS. The job binary must be in a public download link or also stored in the HDFS. It also can enable the QoS monitor and the controller services, described in the submission configuration file (JSON).

### CODE REVIEW

**Comment by igornatanael** [comment]
There are some things that I'd be happy to have here before merge it. 

1 - A client example. This point is mandatory;
2 - A README for the plugin in its main directory, with a brief documentation about the plugin. It is valid for the other plugins but must be and specific merge request for each plugin. This one is less essential, but would be a great advance for our documentation.

File: .gitignore
Comment (tellesnobrega) [inline .gitignore]: why change this?

File: application_manager/plugins/spark_mesos/plugin.py
Comment (igornatanael) [inline application_manager/plugins/spark_mesos/plugin.py]: Remove this line and add a proper LOG message

File: application_manager/plugins/spark_mesos/plugin.py
Comment (igornatanael) [inline application_manager/plugins/spark_mesos/plugin.py]: Remove comments here.

File: application_manager/plugins/spark_mesos/plugin.py
Comment (igornatanael) [inline application_manager/plugins/spark_mesos/plugin.py]: Is it necessary to hold the execution until the end?

File: application_manager/plugins/chronos/plugin.py
Comment (tellesnobrega) [inline application_manager/plugins/chronos/plugin.py]: I'm not sure the chronos plugin changes should be merged together with a pull request named Mesos-spark plugin implementation, but I won't block merging on this.

File: application_manager/plugins/spark_mesos/plugin.py
Comment (igornatanael) [inline application_manager/plugins/spark_mesos/plugin.py]: Remove TODOs

File: application_manager/plugins/spark_mesos/plugin.py
Comment (igornatanael) [inline application_manager/plugins/spark_mesos/plugin.py]: Remove TODO e transform in a common comment

File: application_manager/plugins/spark_mesos/plugin.py
Comment (tellesnobrega) [inline application_manager/plugins/spark_mesos/plugin.py]: Do we need this print?

File: application_manager/plugins/spark_mesos/plugin.py
Comment (igornatanael) [inline application_manager/plugins/spark_mesos/plugin.py]: Remove TODO e transform in a common comment

File: application_manager/plugins/spark_mesos/plugin.py
Comment (igornatanael) [inline application_manager/plugins/spark_mesos/plugin.py]: Remove TODO e transform in a common comment

File: application_manager/plugins/spark_mesos/plugin.py
Comment (tellesnobrega) [inline application_manager/plugins/spark_mesos/plugin.py]: shouldn't this be logged?

File: application_manager/plugins/spark_mesos/plugin.py
Comment (tellesnobrega) [inline application_manager/plugins/spark_mesos/plugin.py]: ditto

File: application_manager/plugins/spark_mesos/plugin.py
Comment (igornatanael) [inline application_manager/plugins/spark_mesos/plugin.py]: Remove these commented lines

File: application_manager/plugins/spark_mesos/plugin.py
Comment (tellesnobrega) [inline application_manager/plugins/spark_mesos/plugin.py]: ditto

File: application_manager/plugins/spark_mesos/plugin.py
Comment (tellesnobrega) [inline application_manager/plugins/spark_mesos/plugin.py]: is it still TODO?

File: application_manager/service/api/v10.py
Comment (igornatanael) [inline application_manager/service/api/v10.py]: Is is working well? If affirmative, remove commented lines.

File: application_manager/plugins/spark_mesos/plugin.py
Comment (tellesnobrega) [inline application_manager/plugins/spark_mesos/plugin.py]: Same question here

File: application_manager/plugins/spark_mesos/plugin.py
Comment (tellesnobrega) [inline application_manager/plugins/spark_mesos/plugin.py]: same

File: application_manager/plugins/spark_mesos/plugin.py
Comment (tellesnobrega) [inline application_manager/plugins/spark_mesos/plugin.py]: same here

File: setup.py
Comment (igornatanael) [inline setup.py:47]: I think we could change the plugin name and put something more specific like: spark-mesos-upv

File: application_manager/plugins/spark_mesos/plugin.py
Comment (igornatanael) [inline application_manager/plugins/spark_mesos/plugin.py]: Change the name for something more specific as SparkMesosUPV

File: setup.sh
Comment (igornatanael) [inline setup.sh]: I think it is an accidental removal

File: application_manager/plugins/spark_mesos/plugin.py
Comment (igornatanael) [inline application_manager/plugins/spark_mesos/plugin.py]: Transform this print line in a proper LOG message

File: application_manager/plugins/spark_mesos/plugin.py
Comment (igornatanael) [inline application_manager/plugins/spark_mesos/plugin.py]: Transform this print line in a proper LOG message

File: application_manager/plugins/spark_mesos/plugin.py
Comment (igornatanael) [inline application_manager/plugins/spark_mesos/plugin.py]: Transform this print line in a proper LOG message

File: application_manager/plugins/chronos/plugin.py
Comment (igornatanael) [reply application_manager/plugins/chronos/plugin.py]: I agree with you. I changed it during the changes for mesos, but I'll move this for another pull request.

File: .gitignore
Comment (betobetodev) [reply .gitignore]: I think this change should be ignored.

File: application_manager/plugins/spark_mesos/plugin.py
Comment (betobetodev) [reply application_manager/plugins/spark_mesos/plugin.py]: Fixed.

File: application_manager/plugins/spark_mesos/plugin.py
Comment (betobetodev) [reply application_manager/plugins/spark_mesos/plugin.py]: Fixed.

File: application_manager/plugins/spark_mesos/plugin.py
Comment (betobetodev) [reply application_manager/plugins/spark_mesos/plugin.py]: Fixed.

File: application_manager/plugins/spark_mesos/plugin.py
Comment (betobetodev) [reply application_manager/plugins/spark_mesos/plugin.py]: No. Fixed.

File: application_manager/plugins/spark_mesos/plugin.py
Comment (betobetodev) [reply application_manager/plugins/spark_mesos/plugin.py]: Fixed.

File: application_manager/plugins/spark_mesos/plugin.py
Comment (betobetodev) [reply application_manager/plugins/spark_mesos/plugin.py]: Fixed.

File: application_manager/plugins/spark_mesos/plugin.py
Comment (betobetodev) [reply application_manager/plugins/spark_mesos/plugin.py]: Fixed.

File: application_manager/plugins/spark_mesos/plugin.py
Comment (betobetodev) [reply application_manager/plugins/spark_mesos/plugin.py]: Fixed.

File: tox.ini
Comment (armstrongmsg) [inline tox.ini]: How about this commented line? Should it be removed?

File: setup.sh
Comment (armstrongmsg) [inline setup.sh]: Can you explain this change? What is the problem with installing tox via apt-get?

File: application_manager/plugins/spark_mesos/plugin.py
Comment (armstrongmsg) [inline application_manager/plugins/spark_mesos/plugin.py]: Why do we need this check? Shouldn't "starting_cap" be always present in "data"? If not, do you realize it is assigning no resources to the application?

File: application_manager/plugins/spark_mesos/plugin.py
Comment (armstrongmsg) [inline application_manager/plugins/spark_mesos/plugin.py]: Does stdout.read() raise an exception when wget fails?

File: application_manager/plugins/spark_mesos/plugin.py
Comment (armstrongmsg) [inline application_manager/plugins/spark_mesos/plugin.py]: Does this call return all applications? Or only the running ones?

File: application_manager/plugins/spark_mesos/plugin.py
Comment (armstrongmsg) [inline application_manager/plugins/spark_mesos/plugin.py]: I think this method should be moved to "utils/mesos.py"

File: application_manager/plugins/spark_mesos/plugin.py
Comment (armstrongmsg) [inline application_manager/plugins/spark_mesos/plugin.py]: I think this method should be moved to "utils/mesos.py"

File: application_manager/plugins/spark_mesos/plugin.py
Comment (armstrongmsg) [inline application_manager/plugins/spark_mesos/plugin.py]: Maybe move this to other file, like "utils/ssh_utils.py"

File: application_manager/plugins/spark_mesos/plugin.py
Comment (armstrongmsg) [inline application_manager/plugins/spark_mesos/plugin.py]: Why are you returning a tuple of None objects? What is the caller expecting as result?

File: application_manager/plugins/spark_mesos/plugin.py
Comment (armstrongmsg) [inline application_manager/plugins/spark_mesos/plugin.py]: Is it possible to split this method in submethods like "get_framework", "get_execution_vm_ids", "get_execution_vm_ips", "start_application"?

**Review by armstrongmsg — CHANGES_REQUESTED** [review]

File: tox.ini
Comment (betobetodev) [inline tox.ini]: I agree.

File: setup.sh
Comment (betobetodev) [reply setup.sh]: What do you think we'll only consider the files with relevant changes to the Mesos plugin? There are many files that do not make sense to change. This is an example. @tellesnobrega @armstrongmsg  @iurygregory @igornatanael

File: application_manager/plugins/spark_mesos/plugin.py
Comment (betobetodev) [reply application_manager/plugins/spark_mesos/plugin.py]: I agree.

File: application_manager/plugins/spark_mesos/plugin.py
Comment (betobetodev) [reply application_manager/plugins/spark_mesos/plugin.py]: I agree.

File: application_manager/plugins/spark_mesos/plugin.py
Comment (betobetodev) [reply application_manager/plugins/spark_mesos/plugin.py]: I agree.

File: application_manager/plugins/spark_mesos/plugin.py
Comment (betobetodev) [reply application_manager/plugins/spark_mesos/plugin.py]: The return of 'execute' method of a plugin provider should be the application id and the object of executor.

File: application_manager/plugins/spark_mesos/plugin.py
Comment (betobetodev) [reply application_manager/plugins/spark_mesos/plugin.py]: Yes. This will be done.

File: .gitignore
Comment (betobetodev) [reply .gitignore]: I search about the ".un" extension and the right form is ".un~". It's a file created by vim.

File: setup.sh
Comment (tellesnobrega) [reply setup.sh]: It makes perfect sense to only change things related to mesos on this pull request

File: application_manager/plugins/spark_mesos/plugin.py
Comment (betobetodev) [reply application_manager/plugins/spark_mesos/plugin.py]: I perceive now that this is a semantic problem in Spark-Mesos plugin. To fix without many changes, my suggestion is a 'busy' flag. What do you think?

File: application_manager/plugins/spark_mesos/plugin.py
Comment (betobetodev) [reply application_manager/plugins/spark_mesos/plugin.py]: To simplify, consider that when 'starting_cap' doesn't exists in config file, this is set to 100.

File: application_manager/plugins/spark_mesos/plugin.py
Comment (betobetodev) [reply application_manager/plugins/spark_mesos/plugin.py]: It's an unusual error. Anyway, this is **TODO: Fix possible wget error**

File: tox.ini
Comment (armstrongmsg) [reply tox.ini]: ?

File: application_manager/plugins/spark_mesos/plugin.py
Comment (armstrongmsg) [inline application_manager/plugins/spark_mesos/plugin.py]: I don't think this check makes any sense. You should remove it.

**Review by armstrongmsg — DISMISSED** [review]

File: application_manager/plugins/spark_mesos/plugin.py
Comment (iurygregory) [inline application_manager/plugins/spark_mesos/plugin.py:1]: application_manager/plugins/spark_mesos/plugin.py:23:1: F401 'ID_Generator' imported but unused
from application_manager.utils.ids import ID_Generator
^
application_manager/plugins/spark_mesos/plugin.py:28:1: F401 'json' imported but unused
import json
^
application_manager/plugins/spark_mesos/plugin.py:29:1: F401 'paramiko' imported but unused
import paramiko
^
application_manager/plugins/spark_mesos/plugin.py:73:13: F841 local variable 'actuator' is assigned to but never used
            actuator = str(data['actuator'])
            ^
application_manager/plugins/spark_mesos/plugin.py:75:30: E127 continuation line over-indented for visual indent
                             else int(data['starting_cap'])
                             ^
application_manager/plugins/spark_mesos/plugin.py:77:54: W291 trailing whitespace
            plugin_log.log("%s | Submission id: %s" %
                                                     ^
application_manager/plugins/spark_mesos/plugin.py:84:45: E127 continuation line over-indented for visual indent
                                            api.cluster_username,
                                            ^
application_manager/plugins/spark_mesos/plugin.py:85:45: E127 continuation line over-indented for visual indent
                                            api.cluster_password,
                                            ^
application_manager/plugins/spark_mesos/plugin.py:86:45: E127 continuation line over-indented for visual indent
                                            api.cluster_key_path)
                                            ^
application_manager/plugins/spark_mesos/plugin.py:98:35: E127 continuation line over-indented for visual indent
                                  '--master mesos://%s:%s ' +
                                  ^
application_manager/plugins/spark_mesos/plugin.py:103:35: E127 continuation line over-indented for visual indent
                                  '--master mesos://%s:%s ' +
                                  ^
application_manager/plugins/spark_mesos/plugin.py:113:1: W293 blank line contains whitespace

^
application_manager/plugins/spark_mesos/plugin.py:115:33: E127 continuation line over-indented for visual indent
                                % (time.strftime("%H:%M:%S")))
                                ^
application_manager/plugins/spark_mesos/plugin.py:116:1: W293 blank line contains whitespace

^
application_manager/plugins/spark_mesos/plugin.py:120:33: E127 continuation line over-indented for visual indent
                                % (time.strftime("%H:%M:%S")))
                                ^
application_manager/plugins/spark_mesos/plugin.py:121:1: W293 blank line contains whitespace

^
application_manager/plugins/spark_mesos/plugin.py:124:33: E127 continuation line over-indented for visual indent
                                % (time.strftime("%H:%M:%S")))
                                ^
application_manager/plugins/spark_mesos/plugin.py:139:31: E127 continuation line over-indented for visual indent
                              % (api.one_username,
                              ^
application_manager/plugins/spark_mesos/plugin.py:146:1: W293 blank line contains whitespace

^
application_manager/plugins/spark_mesos/plugin.py:148:60: E127 continuation line over-indented for visual indent
                                                           self.app_id)
                                                           ^
application_manager/plugins/spark_mesos/plugin.py:164:57: E127 continuation line over-indented for visual indent
                                                        api.one_password,
                                                        ^
application_manager/plugins/spark_mesos/plugin.py:165:57: E127 continuation line over-indented for visual indent
                                                        api.one_url)
                                                        ^
application_manager/plugins/spark_mesos/plugin.py:188:80: E501 line too long (80 > 79 characters)
            plugin_log.log("%s | Starting scaler" % (time.strftime("%H:%M:%S")))
                                                                               ^
application_manager/plugins/spark_mesos/plugin.py:206:80: E501 line too long (80 > 79 characters)
            plugin_log.log("%s | Remove binaries" % (time.strftime("%H:%M:%S")))
                                                                               ^
application_manager/plugins/spark_mesos/plugin.py:251:64: E502 the backslash is redundant between brackets
            handling_thread = threading.Thread(target=executor.\
                                                               ^
application_manager/plugins/spark_mesos/plugin.py:259:44: W291 trailing whitespace
        self.running_application = executor


Many pep8 erros

File: application_manager/utils/mesos.py
Comment (iurygregory) [inline application_manager/utils/mesos.py]: application_manager/utils/mesos.py:19:1: E302 expected 2 blank lines, found 1
def extract_vms_ids(output):
^

File: application_manager/utils/mesos.py
Comment (iurygregory) [inline application_manager/utils/mesos.py:38]: application_manager/utils/mesos.py:37:25: F841 local variable 'e' is assigned to but never used
    except Exception as e:
                        ^

Talvez atualizaro except para mostrar a mensagem da excecao mesmo

File: application_manager/utils/ssh.py
Comment (iurygregory) [inline application_manager/utils/ssh.py]: application_manager/utils/ssh.py:20:24: E127 continuation line over-indented for visual indent
                       password=None, key_path=None):

File: manager.cfg.template
Comment (iurygregory) [inline manager.cfg.template:2]: Não seria melhor manter os valores não? ficaria mais fácil para os usuários (talvez)

**Review by iurygregory — CHANGES_REQUESTED** [review]

File: manager.cfg.template
Comment (betobetodev) [reply manager.cfg.template:2]: manager.cfg.example added.

File: application_manager/utils/mesos.py
Comment (betobetodev) [reply application_manager/utils/mesos.py:38]: Print added.

File: application_manager/plugins/spark_mesos/plugin.py
Comment (betobetodev) [reply application_manager/plugins/spark_mesos/plugin.py:1]: Pep8 errors in this file was fixed.

File: setup.sh
Comment (betobetodev) [reply setup.sh]: Done.

File: tox.ini
Comment (betobetodev) [reply tox.ini]: tox.ini was changed to old version.

File: application_manager/plugins/spark_mesos/plugin.py
Comment (robertonscjr) [reply application_manager/plugins/spark_mesos/plugin.py]: Removed.

Comment (betobetodev) [reply application_manager/plugins/spark_mesos/plugin.py]: Removed.

**Review by igornatanael — APPROVED** [review]
It's ok for me.

**Pull request merged**
